Tools/AP_Bootloader/flash_from_sd.cpp

```diff
@@ -0,0 +1,384 @@
+#include "flash_from_sd.h"
+
+#if AP_BOOTLOADER_FLASH_FROM_SD_ENABLED
+
+#include "ch.h"
+#include "ff.h"
+
+#include "md5.h"
+
+#include <string.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include "stm32_util.h"
+
+#include <AP_HAL_ChibiOS/hwdef/common/flash.h>
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include "support.h"
+
+// swiped from support.cpp:
+static const uint8_t *flash_base = (const uint8_t *)(0x08000000 + (FLASH_BOOTLOADER_LOAD_KB + APP_START_OFFSET_KB)*1024U);
+
+
+// taken from AP_Common.cpp as we don't want to compile the AP_Common directory:
+/**
+ * return the numeric value of an ascii hex character
+ * 
+ * @param[in] a Hexadecimal character 
+ * @return  Returns a binary value
+ */
+int16_t char_to_hex(char a)
+{
+    if (a >= 'A' && a <= 'F')
+        return a - 'A' + 10;
+    else if (a >= 'a' && a <= 'f')
+        return a - 'a' + 10;
+    else
+        return a - '0';
+}
+
+#define MAX_IO_SIZE 4096
+uint8_t buffer[MAX_IO_SIZE];
```

Comment:
can this be static? if not then a better namespace would be good
or a class member?

Reply:
Done